### PR TITLE
[fix] Avoid redelivering duplicated messages when batching is enabled

### PR DESCRIPTION
**Pull request by BewareMyPower**
### Motivation

https://github.com/apache/pulsar/pull/18454 fixed the potential message loss when a batched message is redelivered and one single message of the batch is added to the ACK tracker. However, it also leads to a potential message duplication, see the `testConsumerDedup` test modified by #18454.

The root cause is that single messages will still be passed into the `isDuplicated` method in `receiveIndividualMessagesFromBatch`. However, in this case, the `MessageId` is a `BatchedMessageIdImpl`, while the `MessageId` in `lastCumulativeAck` or `pendingIndividualAcks` are `MessageIdImpl` implementations.

### Modifications

Validate the class type in `isDuplicated` and convert a `BatchedMessageIdImpl` to `MessageIdImpl`. Then revert the unnecessary changes in #18454.

`ConsumerRedeliveryTest#testAckNotSent` is added to verify it works.

### TODO

The duplication could still happen when batch index ACK is enabled. Because even after the ACK tracker is flushed, if only parts of a batched message are not acknowledged, the whole batched message would still be redelivered. I will open another PR to fix it.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/BewareMyPower/pulsar/pull/8

### CODE REVIEW

**Review by codelipenghui — APPROVED** [review]

**Review by congbobo184 — COMMENTED** [review]
LGTM! should we add a test
5 messages in one batch and ackIndividual the 3 messages, redeliver, only can receive the last 2 messages in this batch and enableBatchIndex

**Review by eolivelli — APPROVED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/ConsumerRedeliveryTest.java [hunk 1]
```diff
@@ -65,6 +68,19 @@ public Object[][] ackReceiptEnabled() {
...
+    @DataProvider(name = "batchedMessageAck")
+    public Object[][] batchedMessageAck() {
+        // When batch index ack is disabled (by default), only after all single messages were sent would the pending
+        // ACK be added into the ACK tracker.
+        return new Object[][] {
+                // numAcked, batchSize, ack type
+                { 3, 5, CommandAck.AckType.Individual },
+                { 5, 5, CommandAck.AckType.Individual },
+                { 3, 5, CommandAck.AckType.Cumulative },
+                { 5, 5, CommandAck.AckType.Cumulative }
```
Comment (congbobo184) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/api/ConsumerRedeliveryTest.java:80]: 5, 5 test case may not stable

Comment (BewareMyPower) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/api/ConsumerRedeliveryTest.java:80]: I see, I will take a look soon.

**Review by liangyepianzhou — APPROVED** [review]
LGTM, good work, and the root cause of the [flaky test](https://github.com/apache/pulsar/issues/18466) should be this.

**Comment by BewareMyPower** [comment]
> 5 messages in one batch and ackIndividual the 3 messages, redeliver, only can receive the last 2 messages in this batch and enableBatchIndex

@congbobo184 See the TODO in my PR description. I think it's a bug for batch index ACK and it needs some extra work to fix.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/ConsumerRedeliveryTest.java (hunk 1 above)
Comment (BewareMyPower) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/api/ConsumerRedeliveryTest.java:80]: It's fixed now. The root cause is that after https://github.com/apache/pulsar/pull/18486/commits/b0fad4076f5532103e9f688149048d7889376213, a BatchedMessageIdImpl will be compared with a MessageIdImpl, which should be forbidden.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java [hunk 2]
```diff
@@ -656,6 +662,19 @@ public synchronized void reset() {
         flushRequired = false;
     }
 
+    public synchronized int compareTo(MessageId messageId) {
+        // The position of a message in the batch (BatchMessageIdImpl) must precede the batch itself (MessageIdImpl)
+        if (this.messageId instanceof BatchMessageIdImpl && (!(messageId instanceof BatchMessageIdImpl))) {
+            int result = ((BatchMessageIdImpl) this.messageId).toMessageIdImpl().compareTo(messageId);
```
Comment (coderzc) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java:668]: We can transform `messageId` to `BatchMessageIdImpl`, and `batchIndex = Integer.MAX_VALUE`

Comment (BewareMyPower) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java]: I used `MessageIdImpl#messageIdCompare` instead, PTAL again.

**Review by coderzc — APPROVED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/ConsumerRedeliveryTest.java (hunk 1 above)
Comment (BewareMyPower) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/api/ConsumerRedeliveryTest.java:80]: @congbobo184 PTAL again.

**Comment by congbobo184** [comment]
> > 5 messages in one batch and ackIndividual the 3 messages, redeliver, only can receive the last 2 messages in this batch and enableBatchIndex
> 
> @congbobo184 See the TODO in my PR description. I think it's a bug for batch index ACK and it needs some extra work to fix.

sorry, I miss your description. doesn't the batch index ack fix should include in release/2.9.4?

**Comment by BewareMyPower** [comment]
> doesn't the batch index ack fix should include in release/2.9.4?

I think not. It's not a regression. It's a bug from the beginning when batch index ACK was introduced.

**Pull request merged**